server/src/main/java/io/crate/expression/scalar/formatting/DateTimeFormatter.java

```diff
@@ -0,0 +1,503 @@
+                );
+            case ABBREVIATED_DAY_LOWER:
+                return datetime.getDayOfWeek().getDisplayName(TextStyle.SHORT, Locale.ENGLISH).toLowerCase();
+            case DAY_OF_YEAR:
+                return StringUtils.leftpad(
+                    String.valueOf(datetime.getDayOfYear()),
+                    3,
+                    '0'
+                );
+            case DAY_OF_ISO_WEEK_NUMBERING_YEAR:
+                return StringUtils.leftpad(
+                    String.valueOf(
+                        ((datetime.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR) - 1) * 7) + datetime.getDayOfWeek().getValue()),
+                    3,
+                    '0'
+                );
+            case DAY_OF_MONTH:
+                return StringUtils.leftpad(
+                    String.valueOf(datetime.getDayOfMonth()),
+                    2,
+                    '0'
+                );
+            case DAY_OF_WEEK:
+                return String.valueOf((datetime.getDayOfWeek().getValue() % 7) + 1);
+            case ISO_DAY_OF_WEEK:
+                return String.valueOf(datetime.getDayOfWeek().getValue());
+            case WEEK_OF_MONTH:
+                return String.valueOf((datetime.getDayOfMonth() / 7) + 1);
+            case WEEK_NUMBER_OF_YEAR:
+                return StringUtils.leftpad(
+                    String.valueOf(datetime.get(WEEK_OF_YEAR)),
+                    2,
+                    '0'
+                );
+            case WEEK_NUMBER_OF_ISO_YEAR:
+                return StringUtils.leftpad(
+                    String.valueOf(datetime.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR)),
+                    2,
+                    '0'
+                );
+            case CENTURY:
+                return String.valueOf(((datetime.getYear() - 1) / 100) + 1);
+            case JULIAN_DAY:
+                return String.valueOf(datetime.getLong(JulianFields.JULIAN_DAY));
+            case QUARTER:
+                return String.valueOf((datetime.getMonthValue() + 2) / 3);
+            case ROMAN_MONTH_UPPER:
+                return StringUtils.rightpad(
+                    toRoman(datetime.getMonthValue()).toUpperCase(),
+                    4,
+                    ' '
+                );
+            case ROMAN_MONTH_LOWER:
+                return StringUtils.rightpad(
+                    toRoman(datetime.getMonthValue()).toLowerCase(),
+                    4,
+                    ' '
+                );
+            case TIMEZONE_UPPER:
+            case TIMEZONE_LOWER:
+            case TIMEZONE_HOURS:
+            case TIMEZONE_MINUTES:
+            case TIMEZONE_OFFSET_FROM_UTC:
+            default:
+                return "";
+        }
+    }
+
+    private static String toRoman(int number) {
+        int l = ROMAN_NUMERALS.floorKey(number);
+        if (number == l) {
+            return ROMAN_NUMERALS.get(number);
+        }
+        return ROMAN_NUMERALS.get(l) + toRoman(number - l);
```

Comment:
Nice use of the TreeMap.

It's probably not worth to optimize this - but if you feel like it:

I think you could use modulo to avoid the memory overhead of the map and also turn the recursion into a loop that uses a StringBuilder to create the result. That would avoid some allocations.